js/src/common/utils/ItemList.ts

```diff
@@ -81,51 +172,129 @@ export default class ItemList {
+   *
+   * Content values that are already objects will be proxied and have
+   * `itemName` accessible on them. Primitive values will not have the
+   * `itemName` property accessible.
+   *
+   * **NOTE:** Modifying any objects in the final array may also update the
+   * content of the original ItemList.
+   *
+   * @param keepPrimitives Converts item content to objects and sets the
+   * `itemName` property on them.
    */
-  toArray(): any[] {
-    const items: Item[] = [];
+  toArray(keepPrimitives: true): (T extends object ? T & Readonly<{ itemName: string }> : T)[];
 
-    for (const i in this.items) {
-      if (this.items.hasOwnProperty(i) && this.items[i] instanceof Item) {
-        this.items[i].content = Object(this.items[i].content);
+  toArray(keepPrimitives: boolean = false): T[] | (T & Readonly<{ itemName: string }>)[] {
+    const items: Item<T>[] = Object.keys(this._items).map((key, i) => {
+      const item = this._items[key];
 
-        this.items[i].content.itemName = i;
-        items.push(this.items[i]);
-        this.items[i].key = items.length;
+      if (!keepPrimitives || isObject(item.content)) {
+        // Convert content to object, then proxy it
+        return {
+          ...item,
+          content: this.createItemContentProxy(isObject(item.content) ? item.content : Object(item.content), key),
+        };
+      } else {
+        // ...otherwise just return a clone of the item.
+        return { ...item };
       }
-    }
+    });
+
+    return items.sort((a, b) => b.priority - a.priority).map((item) => item.content);
+  }
 
-    return items
-      .sort((a, b) => {
-        if (a.priority === b.priority) {
-          return a.key - b.key;
-        } else if (a.priority > b.priority) {
-          return -1;
+  /**
+   * A read-only map of all keys to their respective items in no particular order.
+   *
+   * We don't allow adding new items to the ItemList via setting new properties,
+   * nor do we allow modifying existing items directly. You should use the
+   * `.add()` and `.replace()` methods respectively instead.
+   *
+   * @example
+   * const items = new ItemList();
+   * items.add('b', 'My cool value', 20);
+   * items.add('a', 'My value', 10);
+   * items.toObject();
+   * // {
+   * //   a: { content: 'My value', priority: 10, itemName: 'a' },
+   * //   b: { content: 'My cool value', priority: 20, itemName: 'b' },
+   * // }
+   */
+  toObject(): DeepReadonly<Record<string, IItemObject<T>>> {
+    const keyItemMap = Object.keys(this._items).reduce((map, key) => {
+      const obj = {
+        content: this.get(key),
+        itemName: key,
+        priority: this.getPriority(key),
+      };
+
+      map[key] = obj;
+
+      return map;
+    }, {} as Record<string, IItemObject<T>>);
+
+    return keyItemMap;
+  }
+
+  protected createItemContentProxy<C extends object>(content: C, key: string): Readonly<C & { itemName: string }> {
```

Comment:
Mixed feelings about making this protected. Could we go private instead, or at least add a JSDoc declaring it as internal?